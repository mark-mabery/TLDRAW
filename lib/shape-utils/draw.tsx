import { v4 as uuid } from 'uuid'
import * as vec from 'utils/vec'
import { DrawShape, ShapeType } from 'types'
import { registerShapeUtils } from './index'
import { intersectPolylineBounds } from 'utils/intersections'
import { boundsContainPolygon } from 'utils/bounds'
import getStroke from 'perfect-freehand'
import {
  getBoundsFromPoints,
  getRotatedCorners,
  getSvgPathFromStroke,
  translateBounds,
} from 'utils/utils'

const pathCache = new WeakMap<DrawShape, string>([])

const draw = registerShapeUtils<DrawShape>({
  boundsCache: new WeakMap([]),

  create(props) {
    return {
      id: uuid(),
      type: ShapeType.Draw,
      isGenerated: false,
      name: 'Draw',
      parentId: 'page0',
      childIndex: 0,
      point: [0, 0],
      points: [[0, 0]],
      rotation: 0,
      isAspectRatioLocked: false,
      isLocked: false,
      isHidden: false,
      ...props,
      style: {
        strokeWidth: 2,
        strokeLinecap: 'round',
        strokeLinejoin: 'round',
        ...props.style,
        fill: props.style.stroke,
      },
    }
  },

  render(shape) {
    const { id, points, style } = shape

    if (!pathCache.has(shape)) {
      if (points.length < 2) {
        const left = [+style.strokeWidth, 0]
        let d: number[][] = []
        for (let i = 0; i < 10; i++) {
          d.push(vec.rotWith(left, [0, 0], i * ((Math.PI * 2) / 8)))
        }
        pathCache.set(shape, getSvgPathFromStroke(d))
      } else {
        pathCache.set(
          shape,
          getSvgPathFromStroke(
            getStroke(points, { size: +style.strokeWidth * 2, thinning: 0.9 })
          )
        )
      }
    }

    return <path id={id} d={pathCache.get(shape)} />
  },

  applyStyles(shape, style) {
    Object.assign(shape.style, style)
    shape.style.fill = shape.style.stroke
    return this
  },

  getBounds(shape) {
    if (!this.boundsCache.has(shape)) {
      const bounds = getBoundsFromPoints(shape.points)
      this.boundsCache.set(shape, bounds)
    }

    return translateBounds(this.boundsCache.get(shape), shape.point)
  },

  getRotatedBounds(shape) {
    return getBoundsFromPoints(
      getRotatedCorners(this.getBounds(shape), shape.rotation)
    )
  },

  getCenter(shape) {
    const bounds = this.getRotatedBounds(shape)
    return [bounds.minX + bounds.width / 2, bounds.minY + bounds.height / 2]
  },

  hitTest(shape, point) {
    let pt = vec.sub(point, shape.point)
    let prev = shape.points[0]

    for (let i = 1; i < shape.points.length; i++) {
      let curr = shape.points[i]
      if (
        vec.distanceToLineSegment(prev, curr, pt) < +shape.style.strokeWidth
      ) {
        return true
      }
      prev = curr
    }

    return false
  },

  hitTestBounds(this, shape, brushBounds) {
    const b = this.getBounds(shape)
    const center = [b.minX + b.width / 2, b.minY + b.height / 2]

    const rotatedCorners = [
      [b.minX, b.minY],
      [b.maxX, b.minY],
      [b.maxX, b.maxY],
      [b.minX, b.maxY],
    ].map((point) => vec.rotWith(point, center, shape.rotation))

    return (
      boundsContainPolygon(brushBounds, rotatedCorners) ||
      intersectPolylineBounds(
        shape.points.map((point) => vec.add(point, shape.point)),
        brushBounds
      ).length > 0
    )
  },

  rotateTo(shape, rotation) {
    shape.rotation = rotation
    // console.log(shape.points.map(([x, y]) => [x, y]))
    // const bounds = this.getBounds(shape)
    // const center = [bounds.width / 2, bounds.height / 2]
    // shape.points = shape.points.map((pt) => vec.rotWith(pt, center, rotation))
    return this
  },

  translateTo(shape, point) {
    shape.point = vec.toPrecision(point)
    return this
  },

  transform(shape, bounds, { initialShape, scaleX, scaleY }) {
    const initialShapeBounds = this.boundsCache.get(initialShape)
    shape.points = initialShape.points.map(([x, y]) => {
      return [
        bounds.width *
          (scaleX < 0
            ? 1 - x / initialShapeBounds.width
            : x / initialShapeBounds.width),
        bounds.height *
          (scaleY < 0
            ? 1 - y / initialShapeBounds.height
            : y / initialShapeBounds.height),
      ]
    })

    const newBounds = getBoundsFromPoints(shape.points)

    shape.point = vec.sub(
      [bounds.minX, bounds.minY],
      [newBounds.minX, newBounds.minY]
    )
    return this
  },

  transformSingle(shape, bounds, info) {
    this.transform(shape, bounds, info)
    return this
  },

  setProperty(shape, prop, value) {
    shape[prop] = value
    return this
  },

  canTransform: true,
  canChangeAspectRatio: true,
})

export default draw
