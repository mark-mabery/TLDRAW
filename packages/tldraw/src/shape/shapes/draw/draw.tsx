import * as React from 'react'
import { SVGContainer, TLBounds, Utils, TLTransformInfo, ShapeUtil } from '@tldraw/core'
import { Vec } from '@tldraw/vec'
import { intersectBoundsBounds, intersectBoundsPolyline } from '@tldraw/intersect'
import { getStrokeOutlinePoints, getStrokePoints } from 'perfect-freehand'
import { defaultStyle, getShapeStyle } from '~shape/shape-styles'
import { DrawShape, DashStyle, TLDrawShapeType, TLDrawToolType, TLDrawMeta } from '~types'
import { EASINGS } from '~state/utils'

const pointsBoundsCache = new WeakMap<DrawShape['points'], TLBounds>([])
const shapeBoundsCache = new Map<string, TLBounds>()
const rotatedCache = new WeakMap<DrawShape, number[][]>([])
const pointCache: Record<string, number[]> = {}

export const Draw = new ShapeUtil<DrawShape, SVGSVGElement, TLDrawMeta>(() => ({
  type: TLDrawShapeType.Draw,

  toolType: TLDrawToolType.Draw,

  defaultProps: {
    id: 'id',
    type: TLDrawShapeType.Draw,
    name: 'Draw',
    parentId: 'page',
    childIndex: 1,
    point: [0, 0],
    points: [],
    rotation: 0,
    style: defaultStyle,
  },

  Component({ shape, meta, events, isEditing }, ref) {
    const { points, style } = shape

    const polygonPathData = React.useMemo(() => {
      return getFillPath(shape)
    }, [points, style.size, isEditing])

    const pathData = React.useMemo(() => {
      return style.dash === DashStyle.Draw
        ? getDrawStrokePathData(shape, isEditing)
        : getSolidStrokePathData(shape, isEditing)
    }, [points, style.size, style.dash, isEditing])

    const styles = getShapeStyle(style, meta.isDarkMode)

    const strokeWidth = styles.strokeWidth

    // For very short lines, draw a point instead of a line
    const bounds = this.getBounds(shape)

    const verySmall = bounds.width < strokeWidth / 2 && bounds.height < strokeWidth / 2

    if (!isEditing && verySmall) {
      const sw = strokeWidth * 1

      return (
        <SVGContainer ref={ref} id={shape.id + '_svg'} {...events}>
          <circle
            r={strokeWidth * 1}
            fill={styles.stroke}
            stroke={styles.stroke}
            strokeWidth={sw}
            pointerEvents="all"
          />
        </SVGContainer>
      )
    }

    const shouldFill =
      style.isFilled &&
      points.length > 3 &&
      Vec.dist(points[0], points[points.length - 1]) < +styles.strokeWidth * 2

    if (shape.style.dash === DashStyle.Draw) {
      return (
        <SVGContainer ref={ref} id={shape.id + '_svg'} {...events}>
          {shouldFill && (
            <path
              d={polygonPathData}
              stroke="none"
              fill={styles.fill}
              strokeLinejoin="round"
              strokeLinecap="round"
              pointerEvents="fill"
            />
          )}
          <path
            d={pathData}
            fill={styles.stroke}
            stroke={styles.stroke}
            strokeWidth={styles.strokeWidth}
            strokeLinejoin="round"
            strokeLinecap="round"
            pointerEvents="all"
          />
        </SVGContainer>
      )
    }

    // For solid, dash and dotted lines, draw a regular stroke path

    const strokeDasharray = {
      [DashStyle.Draw]: 'none',
      [DashStyle.Solid]: `none`,
      [DashStyle.Dotted]: `${strokeWidth / 10} ${strokeWidth * 3}`,
      [DashStyle.Dashed]: `${strokeWidth * 3} ${strokeWidth * 3}`,
    }[style.dash]

    const strokeDashoffset = {
      [DashStyle.Draw]: 'none',
      [DashStyle.Solid]: `none`,
      [DashStyle.Dotted]: `-${strokeWidth / 20}`,
      [DashStyle.Dashed]: `-${strokeWidth}`,
    }[style.dash]

    const sw = 1 + strokeWidth * 2

    return (
      <SVGContainer ref={ref} id={shape.id + '_svg'} {...events}>
        <path
          d={pathData}
          fill={shouldFill ? styles.fill : 'none'}
          stroke="none"
          strokeWidth={Math.min(4, strokeWidth * 2)}
          strokeLinejoin="round"
          strokeLinecap="round"
          pointerEvents={shouldFill ? 'all' : 'stroke'}
        />
        <path
          d={pathData}
          fill="none"
          stroke={styles.stroke}
          strokeWidth={sw}
          strokeDasharray={strokeDasharray}
          strokeDashoffset={strokeDashoffset}
          strokeLinejoin="round"
          strokeLinecap="round"
          pointerEvents="stroke"
        />
      </SVGContainer>
    )
  },

  Indicator({ shape }) {
    const { points } = shape

    const pathData = React.useMemo(() => {
      return getSolidStrokePathData(shape, false)
    }, [points])

    const bounds = this.getBounds(shape)

    const verySmall = bounds.width < 4 && bounds.height < 4

    if (verySmall) {
      return <circle x={bounds.width / 2} y={bounds.height / 2} r={1} />
    }

    return <path d={pathData} />
  },

  getBounds(shape: DrawShape): TLBounds {
    // The goal here is to avoid recalculating the bounds from the
    // points array, which is expensive. However, we still need a
    // new bounds if the point has changed, but we will reuse the
    // previous bounds-from-points result if we can.

    const pointsHaveChanged = !pointsBoundsCache.has(shape.points)
    const pointHasChanged = !(pointCache[shape.id] === shape.point)

    if (pointsHaveChanged) {
      // If the points have changed, then bust the points cache
      const bounds = Utils.getBoundsFromPoints(shape.points)
      pointsBoundsCache.set(shape.points, bounds)
      shapeBoundsCache.set(shape.id, Utils.translateBounds(bounds, shape.point))
      pointCache[shape.id] = shape.point
    } else if (pointHasChanged && !pointsHaveChanged) {
      // If the point have has changed, then bust the point cache
      pointCache[shape.id] = shape.point
      shapeBoundsCache.set(
        shape.id,
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        Utils.translateBounds(pointsBoundsCache.get(shape.points)!, shape.point)
      )
    }

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return shapeBoundsCache.get(shape.id)!
  },

  shouldRender(prev: DrawShape, next: DrawShape): boolean {
    return next.points !== prev.points || next.style !== prev.style
  },

  hitTestBounds(shape: DrawShape, brushBounds: TLBounds): boolean {
    // Test axis-aligned shape
    if (!shape.rotation) {
      const bounds = this.getBounds(shape)

      return (
        Utils.boundsContain(brushBounds, bounds) ||
        ((Utils.boundsContain(bounds, brushBounds) ||
          intersectBoundsBounds(bounds, brushBounds).length > 0) &&
          intersectBoundsPolyline(
            Utils.translateBounds(brushBounds, Vec.neg(shape.point)),
            shape.points
          ).length > 0)
      )
    }

    // Test rotated shape
    const rBounds = this.getRotatedBounds(shape)

    const rotatedBounds = Utils.getFromCache(rotatedCache, shape, () => {
      const c = Utils.getBoundsCenter(Utils.getBoundsFromPoints(shape.points))
      return shape.points.map((pt) => Vec.rotWith(pt, c, shape.rotation || 0))
    })

    return (
      Utils.boundsContain(brushBounds, rBounds) ||
      intersectBoundsPolyline(
        Utils.translateBounds(brushBounds, Vec.neg(shape.point)),
        rotatedBounds
      ).length > 0
    )
  },

  transform(
    shape: DrawShape,
    bounds: TLBounds,
    { initialShape, scaleX, scaleY }: TLTransformInfo<DrawShape>
  ): Partial<DrawShape> {
    const initialShapeBounds = Utils.getFromCache(this.boundsCache, initialShape, () =>
      Utils.getBoundsFromPoints(initialShape.points)
    )

    const points = initialShape.points.map(([x, y, r]) => {
      return [
        bounds.width *
          (scaleX < 0 // * sin?
            ? 1 - x / initialShapeBounds.width
            : x / initialShapeBounds.width),
        bounds.height *
          (scaleY < 0 // * cos?
            ? 1 - y / initialShapeBounds.height
            : y / initialShapeBounds.height),
        r,
      ]
    })

    const newBounds = Utils.getBoundsFromPoints(shape.points)

    const point = Vec.sub([bounds.minX, bounds.minY], [newBounds.minX, newBounds.minY])

    return {
      points,
      point,
    }
  },
}))

/* -------------------------------------------------- */
/*                       Helpers                      */
/* -------------------------------------------------- */

const STREAMLINE = 0.65

const simulatePressureSettings = {
  simulatePressure: true,
}

const realPressureSettings = {
  easing: (t: number) => t * t,
  simulatePressure: false,
  start: { taper: 1 },
  end: { taper: 1 },
}

function getFillPath(shape: DrawShape) {
  const styles = getShapeStyle(shape.style)

  if (shape.points.length < 2) {
    return ''
  }

  return Utils.getSvgPathFromStroke(
    getStrokePoints(shape.points, {
      size: 1 + styles.strokeWidth * 2,
      thinning: 0.85,
      end: { taper: +styles.strokeWidth * 10 },
      start: { taper: +styles.strokeWidth * 10 },
      last: true,
    }).map((pt) => pt.point)
  )
}

function getDrawStrokePoints(shape: DrawShape, isEditing: boolean) {
  return getStrokePoints(shape.points, {
    streamline: STREAMLINE,
    last: !isEditing,
  })
}

/**
 * Get path data for a stroke with the DashStyle.Draw dash style.
 */
function getDrawStrokePathData(shape: DrawShape, isEditing: boolean) {
  const styles = getShapeStyle(shape.style)

  if (shape.points.length < 2) return ''

  const options = shape.points[1][2] === 0.5 ? simulatePressureSettings : realPressureSettings

  const strokePoints = getDrawStrokePoints(shape, isEditing)

  const stroke = getStrokeOutlinePoints(strokePoints, {
    size: 1 + styles.strokeWidth * 1.618,
    thinning: 0.6,
    streamline: STREAMLINE,
    smoothing: 0.5,
    end: { taper: styles.strokeWidth * 10, easing: EASINGS.easeOutQuad },
    easing: (t) => Math.sin((t * Math.PI) / 2),
    ...options,
    last: !isEditing,
  })

  const path = Utils.getSvgPathFromStroke(stroke)

  return path
}

/**
 * Get SVG path data for a shape that has a DashStyle other than DashStyles.Draw.
 */
function getSolidStrokePathData(shape: DrawShape, isEditing: boolean) {
  const { points } = shape

  if (points.length === 0) return 'M 0 0 L 0 0'

  const strokePoints = getDrawStrokePoints(shape, isEditing).map((pt) => pt.point.slice(0, 2))

  const path = Utils.getSvgPathFromStroke(strokePoints, false)

  return path
}
