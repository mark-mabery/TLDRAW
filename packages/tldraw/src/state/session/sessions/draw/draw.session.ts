import { Utils } from '@tldraw/core'
import { Vec } from '@tldraw/vec'
import { Data, Session, SessionType, TLDrawStatus } from '~types'
import { TLDR } from '~state/tldr'

export class DrawSession implements Session {
  static type = SessionType.Draw
  status = TLDrawStatus.Creating
  topLeft: number[]
  origin: number[]
  previous: number[]
  last: number[]
  points: number[][]
  shiftedPoints: number[][] = []
  shapeId: string
  isLocked?: boolean
  lockedDirection?: 'horizontal' | 'vertical'

  constructor(data: Data, point: number[], id: string) {
    this.origin = point
    this.previous = point
    this.last = point
    this.topLeft = point
    this.shapeId = id

    // Add a first point but don't update the shape yet. We'll update
    // when the draw session ends; if the user hasn't added additional
    // points, this single point will be interpreted as a "dot" shape.
    this.points = [[0, 0, point[2] || 0.5]]
  }

  start = () => void null

  update = (data: Data, point: number[], shiftKey: boolean) => {
    const { shapeId } = this

    // Even if we're not locked yet, we base the future locking direction
    // on the first dimension to reach a threshold, or the bigger dimension
    // once one or both dimensions have reached the threshold.
    if (!this.lockedDirection && this.points.length > 1) {
      const bounds = Utils.getBoundsFromPoints(this.points)
      if (bounds.width > 8 || bounds.height > 8) {
        this.lockedDirection = bounds.width > bounds.height ? 'horizontal' : 'vertical'
      }
    }

    // Drawing while holding shift will "lock" the pen to either the
    // x or y axis, depending on the locking direction.
    if (shiftKey) {
      if (!this.isLocked && this.points.length > 2) {
        // If we're locking before knowing what direction we're in, set it
        // early based on the bigger dimension.
        if (!this.lockedDirection) {
          const bounds = Utils.getBoundsFromPoints(this.points)
          this.lockedDirection = bounds.width > bounds.height ? 'horizontal' : 'vertical'
        }

        this.isLocked = true
        // Start locking
        const returning = [...this.last]

        if (this.lockedDirection === 'vertical') {
          returning[0] = 0
        } else {
          returning[1] = 0
        }

        this.previous = returning
        this.points.push(returning.concat(point[2]))
      }
    } else if (this.isLocked) {
      this.isLocked = false
    }

    if (this.isLocked) {
      if (this.lockedDirection === 'vertical') {
        point[0] = this.origin[0]
      } else {
        point[1] = this.origin[1]
      }
    }

    // The new adjusted point
    const newPoint = Vec.round(Vec.sub(point, this.origin)).concat(point[2])

    // Don't add duplicate points.
    if (Vec.isEqual(this.last, newPoint)) return

    // Add the new adjusted point to the points array
    this.points.push(newPoint)

    // The new adjusted point is now the previous adjusted point.
    this.last = newPoint

    // Does the input point create a new top left?
    const prevTopLeft = [...this.topLeft]

    const topLeft = [Math.min(this.topLeft[0], point[0]), Math.min(this.topLeft[1], point[1])]

    const delta = Vec.sub(topLeft, this.origin)

    // Time to shift some points!
    let points: number[][]

    if (prevTopLeft[0] !== topLeft[0] || prevTopLeft[1] !== topLeft[1]) {
      this.topLeft = topLeft
      // If we have a new top left, then we need to iterate through
      // the "unshifted" points array and shift them based on the
      // offset between the new top left and the original top left.

      points = this.points.map((pt) => {
        return Vec.round(Vec.sub(pt, delta)).concat(pt[2])
      })
    } else {
      // If the new top left is the same as the previous top left,
      // we don't need to shift anything: we just shift the new point
      // and add it to the shifted points array.
      points = [
        ...this.shiftedPoints,
        Vec.sub(newPoint, Vec.sub(topLeft, this.origin)).concat(newPoint[2]),
      ]
    }

    this.shiftedPoints = points

    return {
      document: {
        pages: {
          [data.appState.currentPageId]: {
            shapes: {
              [shapeId]: {
                point: this.topLeft,
                points,
              },
            },
          },
        },
        pageStates: {
          [data.appState.currentPageId]: {
            selectedIds: [shapeId],
          },
        },
      },
    }
  }

  cancel = (data: Data) => {
    const { shapeId } = this
    const pageId = data.appState.currentPageId

    return {
      document: {
        pages: {
          [pageId]: {
            shapes: {
              [shapeId]: undefined,
            },
          },
        },
        pageStates: {
          [pageId]: {
            selectedIds: [],
          },
        },
      },
    }
  }

  complete = (data: Data) => {
    const { shapeId } = this
    const pageId = data.appState.currentPageId

    return {
      id: 'create_draw',
      before: {
        document: {
          pages: {
            [pageId]: {
              shapes: {
                [shapeId]: undefined,
              },
            },
          },
          pageStates: {
            [pageId]: {
              selectedIds: [],
            },
          },
        },
      },
      after: {
        document: {
          pages: {
            [pageId]: {
              shapes: {
                [shapeId]: TLDR.getShape(data, shapeId, pageId),
              },
            },
          },
          pageStates: {
            [data.appState.currentPageId]: {
              selectedIds: [],
            },
          },
        },
      },
    }
  }
}
