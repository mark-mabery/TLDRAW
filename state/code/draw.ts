import CodeShape from './index'
import { uniqueId } from 'utils/utils'
import { DrawShape, ShapeProps, ShapeType } from 'types'
import { defaultStyle } from 'state/shape-styles'

/* ----------------- Start Copy Here ---------------- */

export default class Draw extends CodeShape<DrawShape> {
  constructor(props = {} as ShapeProps<DrawShape>) {
    super({
      id: uniqueId(),

      type: ShapeType.Draw,
      isGenerated: false,
      parentId: (window as any).currentPageId,
      name: 'Draw',
      childIndex: 0,
      point: [0, 0],
      points: [],
      rotation: 0,
      isAspectRatioLocked: false,
      isLocked: false,
      isHidden: false,
      ...props,
      style: {
        ...defaultStyle,
        ...props.style,
      },
    })
  }

  /**
   * Add a point to the draw shape's points.
   *
   * ```ts
   * shape.addPoint([100,100])
   * ```
   */
  addPoint(point: number[]): CodeShape<DrawShape> {
    this.utils.setProperty(this.shape, 'points', [...this.points, point])
    return this
  }

  /**
   * The draw shape's points.
   *
   * ```ts
   * const shapePoints = shape.points
   *
   * shape.points = [[0,0], [100,100], [100,200]]
   * ```
   */
  get points(): number[][] {
    return this.shape.points
  }

  set points(points: number[][]) {
    this.utils.setProperty(this.shape, 'points', points)
  }
}
