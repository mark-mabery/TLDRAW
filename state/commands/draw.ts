import Command from './command'
import history from '../history'
import { Data, DrawShape } from 'types'
import { getPage, setSelectedIds } from 'utils/utils'
import { current } from 'immer'

export default function drawCommand(data: Data, id: string): void {
  const restoreShape = getPage(current(data)).shapes[id] as DrawShape

  history.execute(
    data,
    new Command({
      name: 'create_draw_shape',
      category: 'canvas',
      manualSelection: true,
      do(data, initial) {
        if (!initial) {
          getPage(data).shapes[id] = restoreShape
        }

        setSelectedIds(data, [])
      },
      undo(data) {
        setSelectedIds(data, [])
        delete getPage(data).shapes[id]
      },
    })
  )
}
