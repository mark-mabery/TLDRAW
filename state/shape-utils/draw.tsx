import { uniqueId } from 'utils'
import vec from 'utils/vec'
import { DashStyle, DrawShape, ShapeStyles, ShapeType } from 'types'
import { intersectPolylineBounds } from 'utils/intersections'
import getStroke, { getStrokePoints } from 'perfect-freehand'
import {
  getBoundsCenter,
  getBoundsFromPoints,
  getSvgPathFromStroke,
  translateBounds,
  boundsContain,
} from 'utils'
import { defaultStyle, getShapeStyle } from 'state/shape-styles'
import { registerShapeUtils } from './register'

const rotatedCache = new WeakMap<DrawShape, number[][]>([])
const pathCache = new WeakMap<DrawShape['points'], string>([])
const polygonCache = new WeakMap<DrawShape['points'], string>([])

const draw = registerShapeUtils<DrawShape>({
  boundsCache: new WeakMap([]),

  create(props) {
    return {
      id: uniqueId(),

      type: ShapeType.Draw,
      isGenerated: false,
      name: 'Draw',
      parentId: 'page1',
      childIndex: 0,
      point: [0, 0],
      points: [],
      rotation: 0,
      isAspectRatioLocked: false,
      isLocked: false,
      isHidden: false,
      ...props,
      style: {
        ...defaultStyle,
        ...props.style,
      },
    }
  },

  render(shape) {
    const { id, points, style } = shape

    const styles = getShapeStyle(style)

    if (!pathCache.has(points)) {
      renderPath(shape, style)
    }

    if (points.length > 0 && points.length < 3) {
      return (
        <circle id={id} r={+styles.strokeWidth * 0.618} fill={styles.stroke} />
      )
    }

    const shouldFill =
      points.length > 3 &&
      vec.dist(points[0], points[points.length - 1]) < +styles.strokeWidth * 2

    if (shouldFill && !polygonCache.has(points)) {
      renderFill(shape, style)
    }

    return (
      <g id={id}>
        {shouldFill && (
          <path
            d={polygonCache.get(points)}
            fill={styles.fill}
            strokeWidth="0"
            stroke="none"
          />
        )}
        <path d={pathCache.get(points)} fill={styles.stroke} />
      </g>
    )
  },

  getBounds(shape) {
    if (!this.boundsCache.has(shape)) {
      const bounds = getBoundsFromPoints(shape.points)
      this.boundsCache.set(shape, bounds)
    }

    return translateBounds(this.boundsCache.get(shape), shape.point)
  },

  getRotatedBounds(shape) {
    return translateBounds(
      getBoundsFromPoints(shape.points, shape.rotation),
      shape.point
    )
  },

  getCenter(shape) {
    return getBoundsCenter(this.getBounds(shape))
  },

  hitTest(shape, point) {
    const pt = vec.sub(point, shape.point)
    const min = +getShapeStyle(shape.style).strokeWidth
    return shape.points.some(
      (curr, i) =>
        i > 0 && vec.distanceToLineSegment(shape.points[i - 1], curr, pt) < min
    )
  },

  hitTestBounds(this, shape, brushBounds) {
    // Test axis-aligned shape
    if (shape.rotation === 0) {
      return (
        boundsContain(brushBounds, this.getBounds(shape)) ||
        intersectPolylineBounds(
          shape.points,
          translateBounds(brushBounds, vec.neg(shape.point))
        ).length > 0
      )
    }

    // Test rotated shape
    const rBounds = this.getRotatedBounds(shape)

    if (!rotatedCache.has(shape)) {
      const c = getBoundsCenter(getBoundsFromPoints(shape.points))
      rotatedCache.set(
        shape,
        shape.points.map((pt) => vec.rotWith(pt, c, shape.rotation))
      )
    }

    return (
      boundsContain(brushBounds, rBounds) ||
      intersectPolylineBounds(
        rotatedCache.get(shape),
        translateBounds(brushBounds, vec.neg(shape.point))
      ).length > 0
    )
  },

  transform(shape, bounds, { initialShape, scaleX, scaleY }) {
    const initialShapeBounds = this.boundsCache.get(initialShape)
    shape.points = initialShape.points.map(([x, y, r]) => {
      return [
        bounds.width *
          (scaleX < 0 // * sin?
            ? 1 - x / initialShapeBounds.width
            : x / initialShapeBounds.width),
        bounds.height *
          (scaleY < 0 // * cos?
            ? 1 - y / initialShapeBounds.height
            : y / initialShapeBounds.height),
        r,
      ]
    })

    const newBounds = getBoundsFromPoints(shape.points)

    shape.point = vec.sub(
      [bounds.minX, bounds.minY],
      [newBounds.minX, newBounds.minY]
    )
    return this
  },

  applyStyles(shape, style) {
    const styles = { ...shape.style, ...style }
    styles.dash = DashStyle.Solid
    shape.style = styles
    shape.points = [...shape.points]
    return this
  },

  onSessionComplete(shape) {
    const bounds = this.getBounds(shape)

    const [x1, y1] = vec.sub([bounds.minX, bounds.minY], shape.point)

    shape.points = shape.points.map(([x0, y0, p]) => [x0 - x1, y0 - y1, p])

    this.translateTo(shape, vec.add(shape.point, [x1, y1]))

    return this
  },

  canStyleFill: true,
})

export default draw

const simulatePressureSettings = {
  simulatePressure: true,
}

const realPressureSettings = {
  easing: (t: number) => t * t,
  simulatePressure: false,
  start: { taper: 1 },
  end: { taper: 1 },
}

function renderPath(shape: DrawShape, style: ShapeStyles) {
  const styles = getShapeStyle(style)

  if (shape.points.length < 2) {
    pathCache.set(shape.points, '')
    return
  }

  const options =
    shape.points[1][2] === 0.5 ? simulatePressureSettings : realPressureSettings

  const stroke = getStroke(shape.points, {
    size: 1 + +styles.strokeWidth * 2,
    thinning: 0.85,
    end: { taper: +styles.strokeWidth * 20 },
    start: { taper: +styles.strokeWidth * 20 },
    ...options,
  })

  pathCache.set(shape.points, getSvgPathFromStroke(stroke))
}

function renderFill(shape: DrawShape, style: ShapeStyles) {
  const styles = getShapeStyle(style)

  if (shape.points.length < 2) {
    polygonCache.set(shape.points, '')
    return
  }

  return polygonCache.set(
    shape.points,
    getSvgPathFromStroke(
      getStrokePoints(shape.points, {
        size: 1 + +styles.strokeWidth * 2,
        thinning: 0.85,
        end: { taper: +styles.strokeWidth * 20 },
        start: { taper: +styles.strokeWidth * 20 },
      }).map((pt) => pt.point)
    )
  )
}
