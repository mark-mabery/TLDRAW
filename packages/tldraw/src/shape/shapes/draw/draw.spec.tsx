import { Draw } from './draw'

describe('Draw shape', () => {
  it('Creates an instance', () => {
    new Draw()
  })
})
