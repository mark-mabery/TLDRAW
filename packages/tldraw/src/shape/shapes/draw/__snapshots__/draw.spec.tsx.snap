// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draw shape Creates a shape: draw 1`] = `
Object {
  "childIndex": 1,
  "id": "draw",
  "isComplete": false,
  "name": "Draw",
  "parentId": "page",
  "point": Array [
    0,
    0,
  ],
  "points": Array [],
  "rotation": 0,
  "style": Object {
    "color": "Black",
    "dash": "Draw",
    "isFilled": false,
    "size": "Small",
  },
  "type": "draw",
}
`;
